### ISSUE TITLE
Source Code Emphasis

### ISSUE THREAD

**Issue by obilodeau** [asciidoctor/asciidoctor-reveal.js#225]
I would like source code emphasis to be easier to do. This is a big part of how giving a good technical is about.

A few ways we could accomplish that.

## 1) Leveraging reveal.js' / highlightjs' <mark> support

We could use reveal.js' <mark> support with highlight.js. However, when I tested the content arrives already HTML escaped so something needs to be done.

## 2) Leveraging Asciidoctor's Passthroughs

We could also make use of Asciidoctor's [Passthroughs and substitutions](https://asciidoctor.org/docs/user-manual/#passthroughs) and have something like standard bold markup work inside a source block if a special marker is defined.

## 3) Using a reveal.js plugin

This is also worth exploring and supporting: https://github.com/demoneaux/reveal-code-focus/. However it is line-based, I would rather have something more granular.

## 4) Leveraging Asciidoctor's line highlighting

See Dan's comment below.

I made some tests tonight but nothing worked. It's derailing me from the next release thus this ticket.

**Comment by obilodeau** [comment]
Some info: http://wsmoak.net/2015/09/16/presenting-source-code.html

**Comment by mojavelinux** [comment]
Also keep in mind that Asciidoctor supports line highlighting (not source highlighting, but more like a marker) for select lines. While this only allows emphasis at the granularity of the line level, it's still useful.

It's currently supported for coderay and pygments. See https://github.com/asciidoctor/asciidoctor/blob/76504b08d73bf6aaf35435e9345028ebc6fc89aa/lib/asciidoctor/substitutors.rb#L1481-L1483

**Comment by Mogztter** [comment]
Keywords highlighting can be really useful when you want to explain a query language (for instance SQL):

```
SELECT * FROM table WHERE name = 'Olivier'
```

But I wonder how we could mark tokens/keywords without interfering with the programming language... 
Somehow related to inline callouts: https://github.com/asciidoctor/asciidoctor/issues/3037

If we don't want to mark the actual source code we could use an attribute:

```
[source,sql,highlight-word=SELECT,FROM,WHERE]
----
SELECT * FROM table WHERE name = 'Olivier'
----
```

Not sure how what we should do if the word/token is present multiple times:
- highlight all occurrences
- add an optional index to specify which token we should highlight

Or a more fragile solution with characters index ranges:

```
[source,sql,highlight-char=0-6,9-13,20-25]
----
SELECT * FROM table WHERE name = 'Olivier'
----
```

Should we discuss this feature in Asciidoctor core?

**Comment by obilodeau** [comment]
That's a great syntax suggestion! Adding it on top.

> Should we discuss this feature in Asciidoctor core?

I think so. This could apply to good code documentation. That said, maybe there's something specific to presentations about this... I would need to check good docs / blogs with code vs good presentations with code and try to apply one paradigm to the other.

**Comment by obilodeau** [comment]
Note that a lot of source emphasis options were recently merged in #301. See `examples/source-emphasis.adoc`, `examples/source-highlightjs.adoc` and `examples/source-*` in general.

Options 1, 2 and 4 work now. Option 3 can be explored at users' will since we support adding custom reveal.js plugins.

I'm going to close this as resolved. Feel free to re-open or open a new issue refering to this one if you deem necessary.
